metrics-exporter/src/bin/io_engine/collector/pool_stat.rs

```diff
@@ -0,0 +1,198 @@
+        let pool_bytes_written = GaugeVec::new(pool_bytes_written_opts, &["node", "name"])
+            .expect("Unable to create gauge metric type for pool_committed_size");
+        let pool_num_write_ops = GaugeVec::new(pool_num_write_ops_opts, &["node", "name"])
+            .expect("Unable to create gauge metric type for pool_total_size");
+        let pool_read_latency = GaugeVec::new(pool_read_latency_opts, &["node", "name"])
+            .expect("Unable to create gauge metric type for pool_used_size");
+        let pool_write_latency = GaugeVec::new(pool_write_latency_opts, &["node", "name"])
+            .expect("Unable to create gauge metric type for pool_committed_size");
+        // Descriptors for the custom metrics
+        descs.extend(pool_bytes_read.desc().into_iter().cloned());
+        descs.extend(pool_num_read_ops.desc().into_iter().cloned());
+        descs.extend(pool_bytes_written.desc().into_iter().cloned());
+        descs.extend(pool_num_write_ops.desc().into_iter().cloned());
+        descs.extend(pool_read_latency.desc().into_iter().cloned());
+        descs.extend(pool_write_latency.desc().into_iter().cloned());
+
+        Self {
+            pool_bytes_read,
+            pool_num_read_ops,
+            pool_bytes_written,
+            pool_num_write_ops,
+            pool_read_latency,
+            pool_write_latency,
+            descs,
+        }
+    }
+}
+
+impl Collector for PoolIoStatsCollector {
+    fn desc(&self) -> Vec<&prometheus::core::Desc> {
+        self.descs.iter().collect()
+    }
+
+    fn collect(&self) -> Vec<prometheus::proto::MetricFamily> {
+        let mut c = match Cache::get_cache().lock() {
+            Ok(c) => c,
+            Err(error) => {
+                error!(%error,"Error while getting cache resource");
+                return Vec::new();
+            }
+        };
+        let cp = c.deref_mut();
+        let mut metric_family = Vec::with_capacity(6 * cp.pool_mut().pools.capacity());
+        let node_name = match get_node_name() {
+            Ok(name) => name,
+            Err(error) => {
+                error!(?error, "Unable to get node name");
+                return metric_family;
+            }
+        };
+
+        for i in &cp.pool_iostat_mut().pool_stats {
+            let p: &PoolIoStat = i;
+
+            let pool_bytes_read = match self
+                .pool_bytes_read
+                .get_metric_with_label_values(&[node_name.clone().as_str(), p.name().as_str()])
+            {
+                Ok(pool_bytes_read) => pool_bytes_read,
+                Err(error) => {
+                    error!(%error, "Error while creating metrics(pool_bytes_read) with label values");
+                    return metric_family;
+                }
+            };
+            pool_bytes_read.set(p.bytes_read() as f64);
+            let mut x = pool_bytes_read.collect();
+            metric_family.extend(x.pop());
```

Comment:
What does this do?

Reply:
This concats GaugeVec to the metric family.